src/shared/ProducerConsumerQueue.h

```diff
@@ -0,0 +1,113 @@
+/*
+ * This file is part of the CMaNGOS Project. See AUTHORS file for Copyright information
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+ */
+
+#ifndef _PCQ_H
+#define _PCQ_H
+
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+#include <atomic>
+#include <type_traits>
+
+template <typename T>
+class ProducerConsumerQueue
+{
+    private:
+        std::mutex _queueLock;
+        std::queue<T> _queue;
+        std::condition_variable _condition;
+        std::atomic<bool> _shutdown;
+
+    public:
+
+        ProducerConsumerQueue<T>() : _shutdown(false) { }
+
+        void Push(const T& value)
+        {
+            std::lock_guard<std::mutex> lock(_queueLock);
+            _queue.push(std::move(value));
```

Comment:
The std::move() is deceiving here because value is a const T&, so you're unlikely to call the push you think you're calling. Just remove the std::move (remember, that's just a cast, it's not actually moving anything), or enable real move semantics by having a ```Push( T&& value )``` as well.

Also, this class is perhaps worth making non-copyable.